Feature/library autocomplete

juliankrispel opened the pull request:
Hello @remy. Here is the promised pr. Please review.

I hadn't much time to do this so it's likely I didn't follow your conventions. 

A couple of questions:
- Running the jshint task I'm getting a ton of errors, is there something I'm missing? 
- What browsers do you support? So far I hand-tested in late firefox, safari and chrome.
- no unit-tests?
- are you considering using react or something of the likes for jsbin? Just curious, would make it a lot more fun contributing.

![jsbin-libraries](https://cloud.githubusercontent.com/assets/1188186/9514868/c9af687a-4c9d-11e5-8176-54fe92693b7d.gif)

Also: Thanks so much for jsbin, `tis awesome!

:heart:

remy commented:
I've not checked the branch yet to play - but I will do. Here are initial responses to questions:

> Running the jshint task I'm getting a ton of errors, is there something I'm missing?

Yeah, ignore the whole thing. Grunt is *only* and *exclusively* used to build the min file when going to production. Nothing else. It's just a hangover from legacy code.

> What browsers do you support? So far I hand-tested in late firefox, safari and chrome.

IE8+ and latest-1 on all others. Really the (hand) testing is focused on the DOM, because that's where the quirks are. 

> no unit-tests?

Yup. Welcome to legacy 101. I have a blog post about my tests these days, it's a different situation. If you wanted to take on a PR for tests, you'd pretty much win the Internet!

> are you considering using react or something of the likes for jsbin? Just curious, would make it a lot more fun contributing

Not at all. If I were looking at a framework, it would be more likely polymer due to it's alignment to the Shadow DOM. JS Bin is coming up to 7 years old. I'm prepared to bet money that React code wouldn't work in another 7 years, if it even exists still...

But yeah, I definitely put my hands up (and do so publicly) to say that *most* of the client side code is a mess (the cross file globals is particularly fun, as is the `jsbin.panels.panels...` :) )

juliankrispel commented:
@remy 7 years? WHOW! I had no idea. That's pretty darn amazing :D

remy commented on public/js/editors/library.js:
```diff
@@ -2,69 +2,111 @@
-  analytics.library('select', group.libraries[selected[1]].label);
-  insertResources(library.url);
-  if (library.snippet) {
-    insertSnippet(library.snippet);
```
You've lost snippet support in this line.

remy commented on public/js/editors/library.js:
```diff
@@ -2,69 +2,111 @@
+    selectedItem = 0;
   }
-}).on('click', function () {
-  analytics.library('open');
```
Need to restore analytics

remy commented:
Initial testing feedback:

- [ ] If I use the cursor keys beyond the size of the box, the highlighted library isn't visible (i.e. press down cursor a few times until the highlight is missing)
- [ ] Since grouping is now removed, it should probably be ordered alphanumerically (case-insensitive)
- [ ] If the user filters the results using the input, is it worth show the number of matches [ala the help menu](https://cloudup.com/cocgvzWgSAQ)? (I'm unsure if this is the right thing to do)
- [ ] Remove dividing lines between the results
- [ ] Library search box height should match navigation (that it sits in) - it looks like it's shy by 1px.
- [ ] Restore the snippet support (commented on the line that shows it being lost)
- [ ] Hitting esc should put the cursor focus back to the editor (to test: focus the HTML panel editor, click add library, then press esc - focus is lost)
- [ ] Currently breaks processors that rely on a library pre-pop (for example, on jsbin.com change JavaScript to JSX) - it uses the `#library` element here: https://github.com/jsbin/jsbin/blob/master/public/js/processors/processor.js#L204
- [ ] How does this behave on a mobile device?
- [ ] Opening the library dropdown should close other dropdowns (just call `hideOpen()`)
- [ ] Can you remove the build of public/js/embed.min.js from the PR.
- [ ] When the library selection is re-opened, the filter should be cleared (i.e. open, search for "promise", select, re-open library, only promise is visible, but nothing in the input box).

General question:

- Should the dropdown open to a more information panel saying to search for your favourite library, rather than listing right away? The idea there is that the code would look up from the libraries.js file but also make look ups (beyond the scope of this PR) to cdnjs et al.

Otherwise, really solid start. Couple of breakages, but hopefully straight forward to fix or patch up.

juliankrispel commented:
thanks @remy, I'll get to this soon :)

getsetbro commented:
In jsbin if you add a library it adds the dependencies for you - like jquery. Cdnjs api does not tell you the dependencies of a lib
Also the base api call gives you all the libraries - 1400 right now. But it just gives you 1 URL for that library. For example - for bootstrap you get the main JS file. But what you wanted was the CSS file. So you'd need to request the rest of the bootstrap files from cdnjs, but that would be too many - because it includes the min versions and full versions. So you should have to show the user the rest of them and have them pick the ones they want.

remy commented:
Just checking in, how we doing over here?

juliankrispel commented:
Been swamped with work lately, and child care ;). I'm off to SF for 10 days which means I will get some time to work on stuff like this :+1:

juliankrispel commented:
... within the next few days probably

juliankrispel commented:
@remy I see that another PR is related, I might start adressing all this on the plane tomorrow morning ;)

remy commented:
Ha, cheers. Also, child care > (in an inordinate number of magnitudes) PRs!

Cheers.

juliankrispel commented:
@remy I've scanned the codemirror api docs a little but I'm struggling to find a way to actually focus back on the editor. I thought I'd better ask if you have an idea of how this could be done? Would appreciate some help with that.

remy commented:
If you scan jsbin code for something like jsbin.panels.focused.focus() it
might be what you want.

On Fri, 18 Sep 2015 08:47 Julian Krispel-Samsel <notifications@github.com>
wrote:

> @remy <https://github.com/remy> I've scanned the codemirror api docs a
> little but I'm struggling to find a way to actually focus back on the
> editor. I thought I'd better ask if you have an idea of how this could be
> done? Would appreciate some help with that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jsbin/jsbin/pull/2531#issuecomment-141372530>.
>

juliankrispel commented:
What's done:
- [x] the library view now scrolls with the highlight.
- [x] Search results are sorted alphabetically.
- [x] Removed the dividers between options (good call)
- [x] Library search box now alignes with the navigation
- [x] Restored snippet support
- [x] Restore jsx support
- [x] Opening the library dropdown should close dropdowns and vice versa
- [x] Removed the embed.min.js from PR
- [x] reset filter when re-opening library dropdown

What's still to do:
- [ ] show number of results
- [ ] Test on mobile devices
- [ ] Refocus on editor after choosing lib
- [ ] Instead of showing libraries straight away, show a mesage saying "search for your favourite libs" or something like that...

juliankrispel commented:
@remy I tried that and couldn't get it to work :(

juliankrispel commented:
Right, I gotta hop on a plain now, see you laters

![image](https://cloud.githubusercontent.com/assets/1188186/9954740/b23f69ba-5deb-11e5-821f-deb3aecb6b86.png)

abhisekp commented:
@juliankrispel any updates? This would be a really nice addition :smirk_cat:

juliankrispel commented:
Dude, I was sitting behind you the other day. Talked to Danny and was like,
damn I need to finish this. I'll get on this in the coming weeks

On Sat, Aug 6, 2016 at 10:13 PM, Abhisek Pattnaik <notifications@github.com>
wrote:

> @juliankrispel <https://github.com/juliankrispel> any updates? This would
> be a really nice addition 😼
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jsbin/jsbin/pull/2531#issuecomment-238048997>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABIhWhuCMqZCbTJJo960iahNDCJLJNtxks5qdPj7gaJpZM4FzH1b>
> .
>

juliankrispel commented:
Apologies if this is a little creepy, I should have introduced myself :D

On Sat, Aug 6, 2016 at 10:17 PM, Julian Krispel-Samsel <
julian@goodafternoon.co> wrote:

> Dude, I was sitting behind you the other day. Talked to Danny and was
> like, damn I need to finish this. I'll get on this in the coming weeks
>
> On Sat, Aug 6, 2016 at 10:13 PM, Abhisek Pattnaik <
> notifications@github.com> wrote:
>
>> @juliankrispel <https://github.com/juliankrispel> any updates? This
>> would be a really nice addition 😼
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/jsbin/jsbin/pull/2531#issuecomment-238048997>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ABIhWhuCMqZCbTJJo960iahNDCJLJNtxks5qdPj7gaJpZM4FzH1b>
>> .
>>
>
>

juliankrispel commented:
Whoopsey, I mistook you for remy, making this even more awkward

Julian Krispel-Samsel
rainforestqa.com
goodafternoon.co

> On Aug 6, 2016, at 10:13 PM, Abhisek Pattnaik <notifications@github.com> wrote:
> 
> @juliankrispel any updates? This would be a really nice addition 😼
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

abhisekp commented:
Please don't remove grouping.
Grouping helps e.g. check out http://www.jsdelivr.com/?query=lodash

Grouping actually helps for visual clarity in the pile of libraries.

michalczaplinski commented:
hey, I was just thinking of sending a PR with the same enhancements as you @juliankrispel! :) Did you manage to get it finished?
